crates/downloader/src/download_queue.rs

```diff
@@ -0,0 +1,273 @@
+use std::{
+    collections::BTreeMap,
+    future::Future,
+    time::Duration,
+};
+
+use alloy::{
+    eips::BlockNumberOrTag,
+    providers::{
+        Provider,
+        RootProvider,
+    },
+    rpc::types::Block,
+    transports::{
+        http::Http,
+        RpcError,
+        TransportErrorKind,
+    },
+};
+use futures_util::Stream;
+use tokio::{
+    sync::mpsc,
+    task::JoinSet,
+};
+use tokio_stream::wrappers::ReceiverStream;
+
+#[derive(Debug)]
+pub enum GetBlockError {
+    RateLimited,
+    Other(RpcError<TransportErrorKind>),
+}
+
+pub trait GetBlock {
+    fn get_block(
+        &self,
+        block_number: u64,
+    ) -> impl Future<Output = Result<Option<Block>, GetBlockError>> + Send;
+}
+
+impl GetBlock for RootProvider<Http<reqwest::Client>> {
+    async fn get_block(&self, block_number: u64) -> Result<Option<Block>, GetBlockError> {
+        self.get_block_by_number(BlockNumberOrTag::Number(block_number), true)
+            .await
+            .map_err(|err| match err {
+                RpcError::ErrorResp(resp) if resp.code == 429 => {
+                    GetBlockError::RateLimited
+                }
+                other => GetBlockError::Other(other),
+            })
+    }
+}
+
+enum Mode {
+    /// Batch-download blocks until we reach the end of the chain
+    Batch,
+    /// Poll for new blocks
+    Poll,
+}
+impl Mode {
+    /// How many blocks to download concurrently
+    fn concurrency(&self) -> usize {
+        match self {
+            Mode::Batch => 8,
+            Mode::Poll => 1,
+        }
+    }
+}
+
+pub struct DownloadQueue<P> {
+    provider: P,
+    next_da_height: u64,
+    pending: JoinSet<Result<Option<Block>, GetBlockError>>,
+    completed: BTreeMap<u64, Block>,
+}
+
+impl<P: GetBlock + Clone + Send + 'static> DownloadQueue<P> {
```

Comment:
```suggestion
impl<P> DownloadQueue<P> 
where
    P: GetBlock + Clone + Send + 'static
{
```

Reply:
addressed in [be9b784](https://github.com/FuelLabs/network-watchtower/pull/10/commits/be9b784cd56386015f6aed79291e3affbbb092e9)